### PR TITLE
[WIP] Remove ContinueWith closures from System.Data.SqlClient

### LINKED ISSUE #7787: [perf] System.Data.SqlClient allocates lots of closures with ContinueWith
Ran this grep query:

``` bash
grep --include='*.cs' --exclude-dir=tests --exclude-dir=ref -r 'ContinueWith[^(]*(([^,]*)' .
```

It printed out the following results: https://gist.github.com/anonymous/7afb47c97f2f632c7ddf963111ad3136

The closure in `WinHttpHandler` is fixed by #7785, but there appear to be a bunch of them in `System.Data.SqlClient` that could be removed. Opening up a tracking issue to remove them (which I'll do myself).

---

**edit:** Just realized I wasn't picking up on some of them, because most are written in the form:

``` cs
task.ContinueWith(t => ...)
```

rather than

``` cs
task.ContinueWith((t) => ...)
```

so I ran a new query, which seems to pick up more matches. See: https://gist.github.com/anonymous/14aa1dea4fe3a26e42e338c5b30b09b4

### PR DESCRIPTION
**Pull request by jamesqo**
Fixes #7787 by removing all of the `ContinueWith` closures from `System.Data.SqlClient`. Note that currently, this is still a work in progress; I've only handled the first 3 files in [this list](https://gist.github.com/anonymous/14aa1dea4fe3a26e42e338c5b30b09b4) so far, and haven't tested the changes locally yet. Any feedback/code review would be much appreciated.

cc @justinvp

### CODE REVIEW

**Comment by hughbe** [comment]
> haven't tested the code locally

Why?

File: src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionFactory.cs [hunk 1]
```diff
@@ -217,16 +217,26 @@ internal bool TryGetConnection(DbConnection owningConnection, TaskCompletionSour
...
                                 {
-                                    oldConnection.PrepareForReplaceConnection();
-                                    oldConnection.Dispose();
+                                    capturedOldConnection.PrepareForReplaceConnection();
+                                    capturedOldConnection.Dispose();
                                 }
                                 return newConnection;
-                            }, cancellationTokenSource.Token, TaskContinuationOptions.LongRunning, TaskScheduler.Default);
+                            },
+                            Tuple.Create(this, owningConnection, poolGroup, userOptions, oldConnection),
```
Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionFactory.cs]: The state passed to a task is exposed from its AsyncState property.  If this task is returned out to external code, that code will have access to all of this state.  Do you know that this is safe?  Same question for all cases of this.

Comment (stephentoub) [inline src/System.Data.SqlClient/src/System/Data/ProviderBase/DbConnectionFactory.cs]: Also, how did you verify these changes?  To my knowledge there are no tests that run automatically for SqlClient, and you need to run them manually.  Have you don so, @jamesqo?  

And how have you verified that the closures you were trying to remove have actually been removed?

**Comment by stephentoub** [comment]
I'm closing this.  It doesn't compile and hasn't been tested.

**Comment by jamesqo** [comment]
@stephentoub OK. Sorry for opening this prematurely; I'll submit a new PR once I've finished and run tests.

**Pull request closed without merging**